pudb/debugger.py

```diff
@@ -327,6 +327,7 @@ def set_frame_index(self, index):
             return
 
         self.curframe, lineno = self.stack[index]
+        self.curframe_locals = self.curframe.f_locals
```

Comment:
Maybe add a comment explaining what it's good for? TBH, I'm not sure I understand why the proposed changes have the described effect.

Reply:
I might try to dig deeper to figure out exactly what is going on. From what I can tell, as long as you reuse the same locals object, any modifications you make to it are reflected when the frame is run. But if you access `frame.f_locals` again, it recomputes the locals and resets them. I'm unclear exactly how it works, but it does. 

I'm actually unclear why your example at https://github.com/inducer/pudb/issues/26#issuecomment-486882575 didn't work for you. I tested it and even in Python 2, it has the desired effect when you set `i` on the given line.

Reply:
Well, you've nerd-sniped me. Grr. :slightly_smiling_face: 

Here's what I've found so far:

- Accessing `f_locals` winds up here: https://github.com/python/cpython/blob/c450c8c9ed6e420025f39d0e4850a79f8160cdcd/Objects/frameobject.c#L25-L32
- Which ends up overwriting `f_locals`, which is the same object every time you access it, here: https://github.com/python/cpython/blob/v3.11.0/Objects/frameobject.c#L1097
- I'm still a bit hazy on how changes in `f_locals` make their way from there to the "fast" local variable storage. I lost the trail at `PyFrame_LocalsToFast`, in that I wasn't totally able to pin down what causes it to get called.

Some notes:
- Interestingly, it seems that all this code was significantly rewritten for 3.12.
- There also seem to have been some changes here for 3.11. Search for `PyFrame_LocalsToFast` [here](https://github.com/python/cpython/blob/c450c8c9ed6e420025f39d0e4850a79f8160cdcd/Doc/whatsnew/3.11.rst).

Whatever we do here, we should make sure to preserve a summary of it in comments, so that we don't have to rediscover it some other time.

Reply:
I found https://peps.python.org/pep-0558/, which mentions this behavior (footnote 3), although the link there is broken.

Reply:
That PEP is very technical, but I think what happens is:

- locals is designed to be snapshotted. That's why when you do something like
  ```py
  def f():
      x = 1
      locals()['x'] = 2
      print(x)

  f() 
  ```
  it prints `1` instead of `2`. 
- `frame.f_locals` is designed to be a write-through proxy (probably exactly for this feature). 
- What I think happens is that the descriptor for `f_locals` computes this write-through proxy from a snapshot. That means that when you access it, it recomputes it from what it was, and throws away whatever was there before. Since it's a write-through proxy, this effectively resets the locals in the frame. 

But I'm not completely sure about this. Honestly I think a core dev would need to confirm the details. I think I'll just cross-reference this PEP in the comments since that should be a good jumping off point for anyone who is interested in what is going on.